pallets/liquidity-pools-gateway/src/lib.rs

```diff
@@ -525,14 +616,114 @@ pub mod pallet {
 			Ok((address, incoming_msg))
 		}
 
-		/// Sends the message to the `InboundMessageHandler`.
+		fn clear_storages_for_inbound_messages() {
+			let _ = InboundMessages::<T>::clear(u32::MAX, None);
+			let _ = InboundMessageProofCount::<T>::clear(u32::MAX, None);
+		}
+
+		fn get_expected_message_proof_count() -> u32 {
+			T::MultiRouterCount::get() - 1
+		}
+
+		/// Inserts a message and its expected proof count, or increases the
+		/// message proof count for a particular message.
+		fn update_inbound_message_storages(
+			message: T::Message,
+			weight: &mut Weight,
+		) -> Result<(Proof, u32), DispatchError> {
+			match message.get_message_proof() {
+				None => {
+					let message_proof = message
+						.to_message_proof()
+						.get_message_proof()
+						.expect("message proof ensured by 'to_message_proof'");
+
+					match InboundMessages::<T>::try_mutate(message_proof, |storage_entry| {
+						match storage_entry {
+							None => {
+								*storage_entry =
+									Some((message, Self::get_expected_message_proof_count()));
+							}
+							Some((_, expected_proof_count)) => {
+								// We already have a message, in this case we should expect another
+								// set of message proofs.
+								expected_proof_count
+									.ensure_add_assign(Self::get_expected_message_proof_count())?;
+							}
+						};
+
+						Ok(())
+					}) {
+						Ok(_) => {}
+						Err(e) => return Err(e),
+					};
+
+					*weight = weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+					Ok((
+						message_proof,
+						InboundMessageProofCount::<T>::get(message_proof),
+					))
+				}
+				Some(message_proof) => {
+					let message_proof_count =
+						match InboundMessageProofCount::<T>::try_mutate(message_proof, |count| {
+							count.ensure_add_assign(1)?;
+
+							Ok(*count)
+						}) {
+							Ok(r) => r,
+							Err(e) => return Err(e),
+						};
```

Comment:
Am I missing something or can we simplify readability by utilising `?`?:

```suggestion
						match InboundMessageProofCount::<T>::try_mutate(message_proof, |count| {
							count.ensure_add_assign(1)?;

							Ok(*count)
						})?;
```

Reply:
Might drop these fugly `try_mutate`s.